service/src/main/java/org/apache/griffin/core/job/LivyTaskSubmitHelper.java

```diff
@@ -0,0 +1,215 @@
+import static org.apache.griffin.core.util.JsonUtil.toEntity;
+import org.apache.commons.collections.map.HashedMap;
+import org.quartz.JobDetail;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.core.env.Environment;
+import org.springframework.stereotype.Component;
+import org.springframework.web.client.RestTemplate;
+
+@Component
+public class LivyTaskSubmitHelper {
+
+    private static final Logger logger = LoggerFactory
+            .getLogger(LivyTaskSubmitHelper.class);
+
+    private static final String REQUEST_BY_HEADER = "X-Requested-By";
+    private SparkSubmitJob sparkSubmitJob;
+    private ConcurrentMap<Long, Integer> taskAppIdMap = new ConcurrentHashMap<>();
+    // Current number of tasks
+    private AtomicInteger curConcurrentTaskNum = new AtomicInteger(0);
+    private String workerNamePre;
+    private RestTemplate restTemplate = new RestTemplate();
+    // queue for pub or sub
+    private BlockingQueue<JobDetail> queue;
+    public static final int DEFAULT_QUEUE_SIZE = 20000;
+    private static final int SLEEP_TIME = 300;
+    private String uri;
+
+    @Value("${livy.task.max.concurrent.count:20}")
+    private int maxConcurrentTaskCount;
+    @Value("${livy.task.submit.interval.second:3}")
+    private int batchIntervalSecond;
+
+
+    @Autowired
+    private Environment env;
+
+    /**
+     * Initialize related parameters and open consumer threads.
+     */
+    @PostConstruct
+    public void init() {
+        startWorker();
+        uri = env.getProperty("livy.uri");
+        logger.info("Livy uri : {}", uri);
+    }
+
+    public LivyTaskSubmitHelper() {
+        this.workerNamePre = "livy-task-submit-worker";
+    }
+
+    /**
+     * Initialize blocking queues and start consumer threads.
+     */
+    public void startWorker() {
+        queue = new LinkedBlockingQueue<>(DEFAULT_QUEUE_SIZE);
+        BatchTaskWorker worker = new BatchTaskWorker();
+        worker.setDaemon(true);
+        worker.setName(workerNamePre + "-" + worker.getName());
+        worker.start();
+    }
+
+    /**
+     * Put job detail into the queue.
+     * @param jd job detail.
+     */
+    public void addTaskToWaitingQueue(JobDetail jd) throws IOException {
+        if (jd == null) {
+            logger.warn("task is blank, workerNamePre: {}", workerNamePre);
+            return;
+        }
+
+        if (queue.remainingCapacity() <= 0) {
+            logger.warn("task is discard, workerNamePre: {}, task: {}", workerNamePre, jd);
+            sparkSubmitJob.saveJobInstance(null, NOT_FOUND);
+            return;
+        }
+
+        queue.add(jd);
+
+        logger.info("add_task_to_waiting_queue_success, workerNamePre: {}, task: {}",
+                workerNamePre, jd);
+    }
+
+    /**
+     * Consumer thread.
+     */
+    class BatchTaskWorker extends Thread {
```

Comment:
Don't write our own thread, could you use spring task executor to implements this produce-consume-queue?

Reply:
Do you mean using a thread pool and then implementing a producer consumer queue in a thread pool? In this way, each thread will not get the return result of livy, because the thread submits to livy immediately returns the application id, then livy asynchronously submits, we need to release the resources after the livy commits successfully.

Reply:
Yes, this idea is very important, now use SingleThreadExecutor to manage single-threaded tasks, preventing single-threaded hangs.